### ISSUE TITLE
Sample Weights Error

### ISSUE THREAD

**Issue by alexandergromov** [guillermo-navas-palencia/optbinning#64]
An error occurs when `optbinning.OptimalBinning.fit()` is used with `sample_weight` specified and a dataset contains missing and/or special values.

Example (Kaggle Telco Customer Churn Dataset):

```
# %%

import optbinning
import numpy as np
import pandas as pd

# %%

df = pd.read_csv("WA_Fn-UseC_-Telco-Customer-Churn.csv", sep=",", engine="c")

y = df["Churn"].values

mask = y == "Yes"
y[mask] = 1
y[~mask] = 0
y = y.astype(np.int)


# %%

# Add special cases
df['MonthlyCharges'][0:100] = -1
df['MonthlyCharges'][100:200] = -2

# Add missing values
df['MonthlyCharges'][200:400] = np.nan


# %%

print('n_missing:', df['MonthlyCharges'].isna().sum())
print('n_special_cases:', sum(df['MonthlyCharges'] == -1))
print('n_special_cases:', sum(df['MonthlyCharges'] == -2))

# %%

binning = optbinning.OptimalBinning(dtype = "numerical",
                                    prebinning_method='cart',
                                    special_codes= [-1, -2],
                                    verbose = True)

# Fit with random values as weights
binning.fit(x = df['MonthlyCharges'].values,
            y = y,
            sample_weight = list(np.random.rand(len(df))))

```

Then the error occurs:

```
Traceback (most recent call last):
  File "<input>", line 8, in <module>
  File "/Users/alexander/.conda/envs/woe/lib/python3.7/site-packages/optbinning/binning/binning.py", line 528, in fit
    return self._fit(x, y, sample_weight, check_input)
  File "/Users/alexander/.conda/envs/woe/lib/python3.7/site-packages/optbinning/binning/binning.py", line 781, in _fit
    sw_special, sw_others)
  File "/Users/alexander/.conda/envs/woe/lib/python3.7/site-packages/optbinning/binning/binning.py", line 857, in _fit_prebinning
    ).fit(x, y, sample_weight)
  File "/Users/alexander/.conda/envs/woe/lib/python3.7/site-packages/optbinning/binning/prebinning.py", line 117, in fit
    est.fit(x.reshape(-1, 1), y, sample_weight=sample_weight)
  File "/Users/alexander/.conda/envs/woe/lib/python3.7/site-packages/sklearn/tree/_classes.py", line 894, in fit
    X_idx_sorted=X_idx_sorted)
  File "/Users/alexander/.conda/envs/woe/lib/python3.7/site-packages/sklearn/tree/_classes.py", line 288, in fit
    sample_weight = _check_sample_weight(sample_weight, X, DOUBLE)
  File "/Users/alexander/.conda/envs/woe/lib/python3.7/site-packages/sklearn/utils/validation.py", line 1303, in _check_sample_weight
    .format(sample_weight.shape, (n_samples,)))
ValueError: sample_weight.shape == (7043,), expected (6643,)!
```

I assume, that the error can be fixed by updating the code of `_fit` function in `binning.py`. In this function, `self._fit_prebinning` should be called with `sw_clean` at the current position of `sample_weight`:

```
            splits, n_nonevent, n_event = self._fit_prebinning(
                x_clean, y_clean, y_missing, y_special, y_others,
                self.class_weight, sample_weight, sw_clean, sw_missing,
                sw_special, sw_others)
```

In this case `sw_clean` will be used with `x_clean`, `y_clean` in `_fit_prebinning` that was called, in `prebinning` initialisation:

```
        prebinning = PreBinning(method=self.prebinning_method,
                                n_bins=self.max_n_prebins,
                                min_bin_size=min_bin_size,
                                problem_type=self._problem_type,
                                class_weight=class_weight,
                                **self.prebinning_kwargs
                                ).fit(x, y, sample_weight)
```

Also maybe prebinning initialisation with explicitly `x_clean`, `y_clean`, and `sw_clean` should be used.

**Comment by guillermo-navas-palencia** [comment]
Hi @alexandergromov,

Thanks for pointing this out. I am looking into it.

**Comment by guillermo-navas-palencia** [comment]
You are right, I will make the change.

Thank you.

**Comment by tristanlmiller** [comment]
While this issue was resolved, I noticed that the fix is still not in the pip package distribution.  I did find a way to import the version currently on github, but I think users would appreciate if it were more convenient to access.

**Comment by guillermo-navas-palencia** [comment]
Hi @tristanlmiller,

This fix, along with new features, will be included in v 0.9.0, to be released in January 2021.
